src/slot_provider/clock.rs

```diff
@@ -1,88 +1,184 @@
+    let seconds_until_next_slot = next_slot_start.timestamp() - Utc::now().timestamp();
+
+    tokio::time::sleep(tokio::time::Duration::from_secs(
+        seconds_until_next_slot as u64,
+    ))
+    .await;
+}
+
 pub struct SystemClockSlotProvider {
-    num_slots: Option<usize>,
+    counter: AtomicUsize,
+    max_count: Option<usize>,
+    slots: Mutex<VecDeque<Slot>>,
 }
 
 impl SystemClockSlotProvider {
-    pub fn new() -> Self {
-        Self { num_slots: None }
+    fn internal_new(notify: mpsc::Sender<()>, max_count: Option<usize>) -> Arc<Self> {
+        let provider = Arc::new(Self {
+            counter: AtomicUsize::new(0),
+            max_count,
+            slots: Mutex::new(VecDeque::with_capacity(MAX_SLOTS)),
+        });
+
+        log::info!("Starting slot provider with max_count {:?}", max_count);
+
+        let provider_clone = provider.clone();
+        tokio::spawn(async move {
+            log::debug!("Waiting until next slot to start interval stream");
+            wait_until_next_slot().await;
+
+            // TODO: this is probably broken for leap seconds. A slot is sometimes 11s, sometimes
+            // 13s long. Assuming IntervalStream waits a number of real seconds, not unix timestamp
+            // seconds, we'd become misaligned.
+            let mut interval_stream =
+                IntervalStream::new(interval(Duration::from_secs(SLOT_PERIOD_SECONDS.into())));
+
+            while (interval_stream.next().await).is_some() {
+                let slot = Slot::from_timestamp(Utc::now().timestamp() as u64).unwrap();
+
+                log::debug!(
+                    "Interval stream ticked, adding next slot to buffer {}",
+                    slot.number
+                );
+
+                let mut slots = provider_clone.slots.lock().await;
+                // When we have buffered MAX_SLOTS slots, we drop the oldest slot from the buffer.
+                // Slots are picked up LIFO, the oldest slot is unlikely to ever get processed.
+                // TODO: consider adding a lifetime to a slot instead and picking a buffer which
+                // can always accomodate SLOT_LIFETIME / SLOT_PERIOD_SECONDS slots.
+                if slots.len() == MAX_SLOTS {
+                    let oldest_slot = slots
+                        .pop_front()
+                        .expect("Queue to contain slots after checking length");
+                    log::info!(
+                        "Slots buffer full, dropping oldest slot {} from buffer",
+                        oldest_slot.number
+                    );
+                }
+                slots.push_back(slot);
+
+                // Count how many slots have passed.
+                provider_clone.counter.fetch_add(1, Ordering::Relaxed);
+
+                // Notify any listener about the new slot.
+                notify
+                    .try_send(())
+                    .unwrap_or_else(|_| println!("Error sending notification"));
```

Comment:
whoops, I'll fix this to be `log::warn`. When this fails apparently we saw a new slot, pushed a notification, and now, 12s later, the notification inbox if you will still hasn't been cleared out. i.e. we're losing valuable time to process the most recently available slot!

Reply:
Just to be clear, will this fail / trigger the warning if the buffer on the channel is full ? Or what are expected scenarios where this could happen ?

Reply:
This'll change when we nail down the work scheduling design, let's do that first.